### PR TITLE
Have console host not enter command prompt mode when using `Read-Host -Prompt`

### LINKED ISSUE #4066: Read-Host -Prompt accidentally exposes prompt commands meant for prompting for mandatory parameters only
Note: This problem has been known for a long time - for background, see [this SO answer](https://stackoverflow.com/a/35493243/45375).

`Read-Host -Prompt` - at least in the regular console host and in the ISE in PSv3+ - accidentally exposes functionality that only makes sense in the context of prompting a user for missing mandatory parameter values, namely, the so-called _prompt commands_ that start with `!`.

Note:

* Happens only when the `-Prompt` parameter is used.

* The ISE in PSv3 through PSv5.1 accepts `!` as a _literal_ `!`, but the console host complains about an invalid prompt command (see below); even the ISE, however, interprets `!?` as a prompt command, for instance. (Note that the ISE in v2- used its own, GUI prompt for `Read-Host`, and was therefore not affected.)

Steps to reproduce
------------------

Run in a console window / terminal.

```powershell
> Read-Host -Prompt 'hi'
hi: !  # The "!" is entered by the user.


> Read-Host -Prompt 'hi'
hi: !? # The "!?" is entered by the user.
```

Expected behavior
-----------------

```none
!
!?
```

That is, whatever the user entered should be output as-is, with no interpretation whatsoever.

Actual behavior
---------------

```none
"!" cannot be recognized as a valid Prompt command.
No help is available for hi.
```

That is, 

* `!` was interpreted as the start of a prompt command.

* `!?` was interpreted as the command to print the "parameter's" help message, which, of course, only makes sense in the context of prompting for mandatory parameter values.

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta (v6.0.0-beta.2) on macOS 10.12.5
PowerShell Core v6.0.0-beta (v6.0.0-beta.2) on Ubuntu 16.04.1 LTS
PowerShell Core v6.0.0-beta (v6.0.0-beta.2) on Microsoft Windows 10 Pro (64-bit; v10.0.14393)
Windows PowerShell v5.1.15063.413 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

`Read-Host` calls into `$Host.UI.Prompt()`.  However, this method is also used when the host prompts for mandatory parameters that aren't provided.  The method expects to be called when given a `FieldDescription` and if the input starts with `!` it enters `CommandPromptMode`.  In this mode, you can type `!?` to request help, for example.  However this mode is not something you can use via `Read-Host` (only if calling `$Host.UI.Prompt()` directly passing in a well constructed `FieldDescription`).  When using `Read-Host -Prompt`, the cmdlet creates a `FieldDescription` where the name is the prompt and the rest of the properties are empty.

The fix is that if `Label` is empty, we can assume it's being called from `Read-Host` rather than being called to prompt for a mandatory parameter and thus not enter `CommandPromptMode`.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/4066

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

**Comment by PoshChan** [comment]
@SteveL-MSFT, your last commit had 12 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Send-MailMessage DRT Unit Tests.Can send mail message using named parameters with mandatory parameters
```powershell
Expected exactly 'user01@example.com', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Send-MailMessage.Tests.ps1: line 136
136:         $mail.FromAddress | Should -BeExactly $InputObject.From
```
Send-MailMessage DRT Unit Tests.Can send mail message using named parameters with ReplyTo
```powershell
Expected exactly 'user01@example.com', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Send-MailMessage.Tests.ps1: line 136
136:         $mail.FromAddress | Should -BeExactly $InputObject.From
```
Send-MailMessage DRT Unit Tests.Can send mail message using named parameters with multiple To
```powershell
Expected exactly 'user01@example.com', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Send-MailMessage.Tests.ps1: line 136
136:         $mail.FromAddress | Should -BeExactly $InputObject.From
```
Send-MailMessage DRT Unit Tests.Can send mail message using named parameters with multiple Cc
```powershell
Expected exactly 'user01@example.com', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Send-MailMessage.Tests.ps1: line 136
136:         $mail.FromAddress | Should -BeExactly $InputObject.From
```
Send-MailMessage DRT Unit Tests.Can send mail message using named parameters with multiple Bcc
```powershell
Expected exactly 'user01@example.com', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Send-MailMessage.Tests.ps1: line 136
136:         $mail.FromAddress | Should -BeExactly $InputObject.From
```

**Comment by PoshChan** [comment]
@SteveL-MSFT, your last commit had 13 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Send-MailMessage Feature Tests.Default Port 25.Can send mail with free-form email address
```powershell
Expected exactly 'user01@example.com', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 247
247:             $mail.FromAddress | Should -BeExactly "user01@example.com"
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with high priority
```powershell
Expected exactly 'urgent', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 257
257:             $mail.Priority | Should -BeExactly "urgent"
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with HTML content as body
```powershell
Cannot index into a null array.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 269
269:             $mail.Headers["content-type"] | Should -BeLike "text/html*"

```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with UTF8 encoding
```powershell
Expected exactly 1, but got 0.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 283
283:             $mail.MessageParts.Count | Should -BeExactly 1
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with attachments
```powershell
Expected exactly 3, but got 0.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 303
303:             $mail.MessageParts.Count | Should -BeExactly 3
```

**Comment by PoshChan** [comment]
@SteveL-MSFT, your last commit had 13 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Error occurred in test script '/home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Send-MailMessage.Tests.ps1'
```powershell
Cannot retrieve the dynamic parameters for the cmdlet. An exception was thrown by a TaskScheduler.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Send-MailMessage.Tests.ps1: line 11
```
Error occurred in test script '/home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Set-Date.Tests.ps1'
```powershell
Too many open files
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Set-Date.Tests.ps1: line 4
```
Error occurred in test script '/home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Send-MailMessage.Tests.ps1'
```powershell
Cannot retrieve the dynamic parameters for the cmdlet. An exception was thrown by a TaskScheduler.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Send-MailMessage.Tests.ps1: line 11
```
Error occurred in test script '/home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Set-Date.Tests.ps1'
```powershell
Too many open files
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Set-Date.Tests.ps1: line 4
```
Invoke-WebRequest tests.Error occurred in Describe block
```powershell
The 'Start-WebListener' command was found in the module 'WebListener', but the module could not be loaded. For more information, run 'Import-Module WebListener'.
At /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:372 char:24
```

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry all

**Comment by PoshChan** [comment]
@SteveL-MSFT, successfully started retry of `PowerShell-CI-static-analysis, PowerShell-CI-Windows, PowerShell-CI-macOS, PowerShell-CI-Linux`

**Comment by PoshChan** [comment]
@SteveL-MSFT, your last commit had 13 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Send-MailMessage Feature Tests.Default Port 25.Can send mail with free-form email address
```powershell
Expected exactly 'user01@example.com', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 247
247:             $mail.FromAddress | Should -BeExactly "user01@example.com"
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with high priority
```powershell
Expected exactly 'urgent', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 257
257:             $mail.Priority | Should -BeExactly "urgent"
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with HTML content as body
```powershell
Cannot index into a null array.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 269
269:             $mail.Headers["content-type"] | Should -BeLike "text/html*"

```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with UTF8 encoding
```powershell
Expected exactly 1, but got 0.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 283
283:             $mail.MessageParts.Count | Should -BeExactly 1
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with attachments
```powershell
Expected exactly 3, but got 0.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 303
303:             $mail.MessageParts.Count | Should -BeExactly 3
```

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry windows

**Comment by PoshChan** [comment]
@SteveL-MSFT, successfully started retry of `PowerShell-CI-Windows`

**Comment by PoshChan** [comment]
@SteveL-MSFT, your last commit had 13 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Send-MailMessage Feature Tests.Default Port 25.Can send mail with free-form email address
```powershell
Expected exactly 'user01@example.com', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 247
247:             $mail.FromAddress | Should -BeExactly "user01@example.com"
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with high priority
```powershell
Expected exactly 'urgent', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 257
257:             $mail.Priority | Should -BeExactly "urgent"
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with HTML content as body
```powershell
Cannot index into a null array.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 269
269:             $mail.Headers["content-type"] | Should -BeLike "text/html*"

```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with UTF8 encoding
```powershell
Expected exactly 1, but got 0.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 283
283:             $mail.MessageParts.Count | Should -BeExactly 1
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with attachments
```powershell
Expected exactly 3, but got 0.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 303
303:             $mail.MessageParts.Count | Should -BeExactly 3
```

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry windows

**Comment by PoshChan** [comment]
@SteveL-MSFT, successfully started retry of `PowerShell-CI-Windows`

**Comment by PoshChan** [comment]
@SteveL-MSFT, your last commit had 13 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Send-MailMessage Feature Tests.Default Port 25.Can send mail with free-form email address
```powershell
Expected exactly 'user01@example.com', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 247
247:             $mail.FromAddress | Should -BeExactly "user01@example.com"
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with high priority
```powershell
Expected exactly 'urgent', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 257
257:             $mail.Priority | Should -BeExactly "urgent"
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with HTML content as body
```powershell
Cannot index into a null array.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 269
269:             $mail.Headers["content-type"] | Should -BeLike "text/html*"

```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with UTF8 encoding
```powershell
Expected exactly 1, but got 0.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 283
283:             $mail.MessageParts.Count | Should -BeExactly 1
```
Send-MailMessage Feature Tests.Default Port 25.Can send mail with attachments
```powershell
Expected exactly 3, but got 0.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Send-MailMessage.Tests.ps1: line 303
303:             $mail.MessageParts.Count | Should -BeExactly 3
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Read-Host.Tests.ps1 [hunk 1]
```diff
@@ -9,28 +9,37 @@ Describe "Read-Host Test" -tag "CI" {
...
+
     It "Read-Host returns a secure string when using -AsSecureString parameter" {
         $result = $ps.AddScript("Read-Host -AsSecureString").Invoke() | select-object -first 1
         $result | Should -BeOfType SecureString
-        [pscredential]::New("foo",$result).GetNEtworkCredential().Password | Should -BeExactly TEST
+        [pscredential]::New("foo",$result).GetNetworkCredential().Password | Should -BeExactly TEST
+    }
 
+    It "Read-Host doesn't enter command prompt mode" {
+        $result = "!1" | pwsh -nop -c "Read-host -Prompt 'foo'"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Read-Host.Tests.ps1]:
```suggestion
        $result = "!1" | pwsh -NoProfile -c "Read-host -Prompt 'foo'"
```

**Review by iSazonov — APPROVED** [review]
LGTM with one minor comment.

**Comment by SteveL-MSFT** [comment]
@PoshChan please remind me in 1 minute

**Comment by PoshChan** [comment]
@SteveL-MSFT, this is the reminder you requested 1 minute ago

**Comment by PoshChan** [comment]
@SteveL-MSFT, your last commit had 1 failures in `PowerShell-CI-windows`
Read-Host Test.Read-Host doesn't enter command prompt mode
```powershell
Expected exactly @('foo: !1', '!1'), but got @('!1', '!1').
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Read-Host.Tests.ps1: line 43
43:         $result | should -BeExactly @('foo: !1','!1')
```

**Comment by SteveL-MSFT** [comment]
Looks like the failure is that on Windows, the prompt is written directly to the console but on non-Windows it's written to stdout.  Seems like writing to stdout is the right thing here.

**Comment by SteveL-MSFT** [comment]
Looks like on Windows, it relies on native console apis due to line input mode, so too big of a change to have it use .Net Console class.

**Comment by iSazonov** [comment]
Perhaps we could use our test module HelpersHostCS  on Windows.

**Comment by SteveL-MSFT** [comment]
@iSazonov this is a change in the console host so a test host doesn't help here.

**Comment by daxian-dbw** [comment]
`Read-Host` calls into `$Host.UI.Prompt()`.  However, this method is also used when the host prompts for mandatory parameters that aren't provided.  The method expects to be called when given a `FieldDescription` and if the input starts with `!` it enters `CommandPromptMode`.  In this mode, you can type `!?` to request help, for example.  However this mode is not something you can use via `Read-Host` (only if calling `$Host.UI.Prompt()` directly passing in a well constructed `FieldDescription`).  When using `Read-Host -Prompt`, the cmdlet creates a `FieldDescription` where the name is the prompt and the rest of the properties are empty.

The fix is that if `Label` is empty, we can assume it's being called from `Read-Host` rather than being called to prompt for a mandatory parameter and thus not enter `CommandPromptMode`.

**Pull request merged**